src/main/java/org/zalando/nakadi/repository/kafka/KafkaTopicRepository.java

```diff
@@ -164,43 +165,89 @@ public boolean partitionExists(final String topic, final String partition) throw
+            Thread.currentThread().interrupt();
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        } catch (final RuntimeException e) {
+            item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
+            throw new EventPublishingException("Error publishing message to kafka", e);
+        }
+    }
 
-            try {
-                kafkaProducer.send(record, kafkaSendCallback(item, done));
-            } catch (InterruptException | SerializationException | BufferExhaustedException e) {
-                item.updateStatusAndDetail(EventPublishingStatus.FAILED, "internal error");
-                throw new EventPublishingException("Error publishing message to kafka", e);
-            }
+    private static boolean isExceptionShouldLeadToReset(
+            @Nullable final Exception exception) {
+        if (null == exception) {
+            return false;
         }
+        return Stream.of(NotLeaderForPartitionException.class, UnknownTopicOrPartitionException.class).
+                filter(clazz -> clazz.isAssignableFrom(exception.getClass()))
+                .findAny().isPresent();
+    }
 
+    @Override
+    public void syncPostBatch(final String topicId, final List<BatchItem> batch) throws EventPublishingException {
+        batch.forEach(item -> Preconditions.checkNotNull(
+                item.getPartition(), "BatchItem partition can't be null at the moment of publishing!"));
+        final Producer<String, String> producer = kafkaFactory.takeProducer();
+        final Map<BatchItem, CompletableFuture<Exception>> sendFutures = new HashMap<>();
         try {
-            final boolean isSuccessful = done.await(createSendTimeout(), TimeUnit.MILLISECONDS);
-
-            if (!isSuccessful) {
-                failUnpublished(batch, "timed out");
-                throw new EventPublishingException("Timeout publishing events");
+            for (final BatchItem item : batch) {
+                item.setStep(EventPublishingStep.PUBLISHING);
+                sendFutures.put(item, publishItem(producer, topicId, item));
             }
-
-            final boolean atLeastOneFailed = batch.stream()
-                    .anyMatch(item -> item.getResponse().getPublishingStatus() == EventPublishingStatus.FAILED);
-            if (atLeastOneFailed) {
-                failUnpublished(batch, "internal error");
-                throw new EventPublishingException("Error publishing message to kafka");
+            final CompletableFuture<Void> multiFuture = CompletableFuture.allOf(
+                    sendFutures.values().toArray(new CompletableFuture<?>[sendFutures.size()]));
+            multiFuture.get(createSendTimeout(), TimeUnit.MILLISECONDS);
+
+            // Now lets check for errors
+            final boolean needReset = sendFutures.entrySet().stream().filter(
```

Comment:
For me this one is cleaner:
            `sendFutures.entrySet().stream().filter(
                    entry -> isExceptionShouldLeadToReset(entry.getValue().getNow(null)))
                    .findAny()
                    .ifPresent(entry -> kafkaFactory.terminateProducer(producer));`